### PR TITLE
Backport #25415: fips: RNG health tests for FIPS 140-3

### PR DESCRIPTION
**Pull request by paulidale**
Backport of #25415 to 3.4

FIPS 140-3 mandates that the health tests in SP 800-90B section 4 are done on the output of any entropy source.

They are implemented here as a replacement for the primary DRBG in the FIPS provider. This results in a setup that looks like this:
```
               +-------------+
               |             |
               | Seed Source |
               |             |
               +------+------+
                      |
                      |
                      v
               +-------------+
               |             |
               |  CRNG Test  |
               |             |
               ++----------+-+
                |          |
                |          |
                v          v
    +--------------+     +--------------+
    |              |     |              |
    | Public DRBG  |     | Private DRBG |
    |              |     |              |
    +--------------+     +--------------+
```

An additional benefit, that of avoiding DRBG chains, is also gained. The current standards do not permit the output of one DRBG to be used as the input for a second (i.e. a chain).

This also leaves open the future possibility of incorporating a seed source inside the FIPS boundary.
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Rebased to remove fairly trivial conflict from #25511.

File: doc/man7/EVP_RAND-CRNG-TEST.pod [hunk 1]
```diff
@@ -0,0 +1,72 @@
+=pod
+
+=head1 NAME
+
+EVP_RAND-CRNG-TEST - The FIPS health testing EVP_RAND filter
+
+=head1 DESCRIPTION
+
+This B<EVP_RAND> object acts a filter between the entropy source
```
Comment (slontis) [inline doc/man7/EVP_RAND-CRNG-TEST.pod:9]: as a filter

Comment (paulidale) [reply doc/man7/EVP_RAND-CRNG-TEST.pod]: Fixed.

File: doc/man7/EVP_RAND-CRNG-TEST.pod [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ EVP_RAND *rand = EVP_RAND_fetch(NULL, "CRNG-TEST", NULL);
+ EVP_RAND_CTX *rctx = EVP_RAND_CTX_new(rand, parent);
+
+=head1 SEE ALSO
+
+L<EVP_RAND(3)>, L<OSSL_PROVIDER-FIPS(7)>
+
+=head1 HISTORY
+
+This functionality was added in OpenSSL 3.5.
```
Comment (slontis) [inline doc/man7/EVP_RAND-CRNG-TEST.pod:61]: 3.5??

**Review by slontis — DISMISSED** [review]

**Comment by t8m** [comment]
LGTM but could you please add a CHANGES.md entry? Also please create a new PR against the master branch which will sync the fixes missing there.

**Comment by jvdsn** [comment]
This change is quite strange to me. Section 4 of SP 800-90B explicitly says:
> Health tests are applied to the outputs of a noise source before any conditioning is done. (It is permissible to also apply some health tests to conditioned outputs, but this is not required.)

As far as I know, OpenSSL does not implement a "noise source" but instead gets entropy from other entropy sources (such as getrandom) which already perform conditioning?

**Review by t8m — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the 3.4 branch. Thank you for your contribution.

**Pull request closed without merging**